bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+    }))
+}
+
+/// Free a previously created runtime.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_free(runtime: *mut tmprl_runtime_t) {
+    if !runtime.is_null() {
+        unsafe {
+            Box::from_raw(runtime);
+        }
+    }
+}
+
+/// A core instance owned by Core. This must be passed to tmprl_core_shutdown
+/// when no longer in use which will free the resources.
+pub struct tmprl_core_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+    // TODO(cretz): Any concerns with the dynamic dispatch overhead here?
+    // TODO(cretz): Should we ask Rust SDK to give us back a struct instead?
+    // TODO(cretz): We could use generics, but impl traits don't cross the FFI
+    // boundary properly
+    core: std::sync::Arc<dyn temporal_sdk_core_api::Core>,
+}
+
+/// Callback called by tmprl_core_init on completion. The first parameter of the
+/// callback is user data passed into the original function. The second
+/// parameter is a core instance if the call is successful or null if not. If
+/// present, the core instance must be freed via tmprl_core_shutdown when no
+/// longer in use. The third parameter of the callback is a byte array for a
+/// InitResponse protobuf message which must be freed via tmprl_bytes_free.
+type tmprl_core_init_callback = unsafe extern "C" fn(
+    user_data: *mut libc::c_void,
+    core: *mut tmprl_core_t,
+    resp: *const tmprl_bytes_t,
+);
+
+// TODO(cretz): Catch unwind for badly behaved callbacks?
+/// Callback called on function completion. The first parameter of the callback
+/// is user data passed into the original function. The second parameter of the
+/// callback is a never-null byte array for a response protobuf message which
+/// must be freed via tmprl_bytes_free.
+type tmprl_callback = unsafe extern "C" fn(user_data: *mut libc::c_void, core: *const tmprl_bytes_t);
+
+/// Create a new core instance.
+///
+/// The runtime is required and must outlive this instance. The req_proto and
+/// req_proto_len represent a byte array for a InitRequest protobuf message. The
+/// callback is invoked on completion.
+#[no_mangle]
+pub extern "C" fn tmprl_core_init(
+    runtime: *mut tmprl_runtime_t,
+    req_proto: *const u8,
+    req_proto_len: libc::size_t,
+    user_data: *mut libc::c_void,
+    callback: tmprl_core_init_callback,
+) {
+    let runtime = unsafe { &*runtime };
+    let req = match tmprl_core_t::decode_proto::<bridge::InitRequest>(req_proto, req_proto_len) {
+        Ok(req) => req,
+        Err(message) => {
+            let resp = bridge::InitResponse {
+                error: Some(bridge::init_response::Error { message }),
+            };
+            unsafe {
+                callback(
+                    user_data,
+                    std::ptr::null_mut(),
+                    tmprl_bytes_t::from_vec(resp.encode_to_vec()).into_raw(),
+                );
+            }
+            return;
+        }
+    };
+    let user_data = Box::new(UserDataHandle(user_data));
```

Comment:
Is this box for deallocating the memory after the task runs? An explicit `drop` inside the task at the end might make that more clear

Reply:
Looks like this pattern is in a few places

Reply:
The box and the `UserDataHandle` is only for raw pointer access across threads. See https://doc.rust-lang.org/nomicon/send-and-sync.html.

I'd have to recheck the compiler error on why I needed to box this struct, but for whatever reason, just marking it as unsafe send/sync was not enough..

Reply:
After research, I can remove the box alloc and just reference the handle

Reply:
Done